pandas/core/groupby/ops.py

```diff
@@ -825,7 +827,9 @@ def _is_indexed_like(obj, axes) -> bool:
 
 
 class DataSplitter:
-    def __init__(self, data, labels, ngroups, axis: int = 0):
+    fast_apply = None
+
+    def __init__(self, data: NDFrame, labels, ngroups: int, axis: int = 0):
```

Comment:
I *think* this is a generic implementation, so should use `FrameOrSeries` instead of `NDFrame` (may help some of the other issues from your OP)

Reply:
I just tried changing all the `NDFrame` annotations to `FrameOrSeries` and now I'm getting additional complaints:

```
pandas/core/groupby/ops.py:869: error: Return type "Series" of "_chop" incompatible with return type "FrameOrSeries" in supertype "DataSplitter"
pandas/core/groupby/ops.py:881: error: Return type "DataFrame" of "_chop" incompatible with return type "FrameOrSeries" in supertype "DataSplitter"
```

Was this comment supposed to just apply to the `__init__` on 832?  Regardless, WTF?

Reply:
I was wrong before; I guess as a rule of thumb if it gets subclassed OK to type as NDFrame in superclass and DataFrame / Series in appropriate subclass. However, if it doesn't get subclassed it's probably a generic function and should utilize FrameOrSeries